src/cryptonote_core/blockchain.cpp

```diff
@@ -2571,248 +2548,200 @@ bool Blockchain::check_tx_inputs(transaction& tx, tx_verification_context &tvc,
 {
   PERF_TIMER(check_tx_inputs);
   LOG_PRINT_L3("Blockchain::" << __func__);
-  size_t sig_index = 0;
   if(pmax_used_block_height)
     *pmax_used_block_height = 0;
 
-  crypto::hash tx_prefix_hash = get_transaction_prefix_hash(tx);
-
   const uint8_t hf_version = m_hardfork->get_current_version();
 
-  if (hf_version >= 9 && tx.version < 2)
+  // Min/Max Version Check
   {
-    tvc.m_invalid_version = true;
-    return false;
-  }
-  else if (hf_version < 9 && tx.version > 2)
-  {
-    tvc.m_invalid_version = true;
-    return false;
+    if      (hf_version >= network_version_10_bulletproofs) tvc.m_invalid_version = (tx.version <  transaction::version_3_per_output_unlock_times);
+    else if (hf_version >= network_version_9_service_nodes) tvc.m_invalid_version = (tx.version <  transaction::version_2);
+    else                                                    tvc.m_invalid_version = (tx.version != transaction::version_2);
+
+    if (tvc.m_invalid_version)
+      return false;
   }
 
-  // from hard fork 2, we require mixin at least 2 unless one output cannot mix with 2 others
-  // if one output cannot mix with 2 others, we accept at most 1 output that can mix
-  if (hf_version >= 2 && !tx.is_deregister_tx())
+  if (tx.is_deregister_tx())
```

Comment:
I wonder if this should be taken out into a separate function

Reply:
Reasons for? There's not much stack setup leaking into scopes that don't need it and not a DRY situation.

Reply:
I'm not if favour of methods that are 350+ lines long. But I had another look and realised that the method is basically one large if-else statement, which is easy enough to read, so it is OK.